docs/cli/error_handling.md

```diff
@@ -0,0 +1,166 @@
+# Error handling principles for the CLI
+
+* These follow from the [General Principles](https://vault.shopify.io/teams/734/pages/Error-Handling-Principles~zEv1.md) document.
+
+## Global type and handler for fatal errors
+
+You’ll see in the section below that the CLI currently makes an exception to some of the general principles presented in this doc. This may not be true forever since there are some upcoming projects that will change the tradeoffs.
+
+### Leverage the `FatalError` hierarchy to short circuit the program and present a custom UX to the user
+
+The CLI defines `FatalError` and several subtypes of `FatalError`. Within the CLI codebase, you can raise any of these errors at any time to **fully halt the program and present a specific UX to the user**.
+
+**Principle**: choose the right error type if you need to abort the program and handle the situation in a specific way:
+
+* Use `AbortError` for expected scenarios requiring user action
+* Use `BugError` for programmer errors and invariant violations
+* Use `AbortSilentError` for user-initiated cancellations
+* Use `ExternalError` when external commands fail
+
+#### AbortError
+
+Represents expected user/environment issues that require user action
+
+**Characteristics:**
+
+* Most common error type in the CLI
+* Provides UI guidance to users
+* Reported as "expected\_error" to analytics (it’s a user problem, not a Shopify one)
+* Sent to Bugsnag as "handled"
+
+#### BugError
+
+Represents unexpected issues that indicate CLI bugs
+
+**Characteristics:**
+
+* Always reported as "unexpected\_error" and sent to Bugsnag as "unhandled"
+* Shows stack trace in development
+* Should be used for invariant violations and programmer errors
+
+#### AbortSilentError
+
+Silent termination without any output
+
+**Characteristics:**
+
+* Used when user cancels operations (e.g., declining migrations)
+* No message rendered to user
+* Clean exit without error indicators
+
+#### ExternalError
+
+Errors from external command execution
+
+**Characteristics:**
+
+* Extends AbortError with command context
+* Shows which external command failed
+* Helps users understand issues with dependencies
+
+### Divergence from general principles {#divergence-from-general-principles}
```

Comment:
This section and the next one (the future) maybe fit better in the decision-record folder. It's a bit different than the rest of the file that talks about actual guidelines.